### PR TITLE
Module deprecator

### PR DESCRIPTION
**Pull request by balopat**
Introduces deprecated module reference. 

We are coming up to a couple of module refactors, e.g. `cirq.experiments` -> `cirq.qcvv` and `cirq.google` -> `cirq_google`.

With the functionality introduced here, we can achieve backwards compatible reference by adding a similar snippet to the one below to the original parent module's `__init__.py` file: 

```
from cirq import _compat 
_compat.deprecated_submodule(
    new_module_name="cirq.experiments",
    old_parent="cirq",
    old_child="qcvv",
    deadline="v0.20",
    create_attribute=True,
)
```

Callers of all types will get a single deprecation warning. 

Design doc: https://tinyurl.com/cirq-deep-aliasing

Extensive testing was done on the full `cirq.google` extraction - all the interesting edge cases  were captured as unit tests here.

Disclaimers: 
  - the tests need to run in a separate subprocess to ensure that the import system is unaffected (I tried resetting sys.modules, and all the other entry points as a pytest fixture, but it didn't really work due to some side-effects I couldn't figure out) - thus I rolled a mechanism that plays nicely within pytest, using multiprocessing.
  - there is a g3 specific problem with par files as there is a specific Loader used internally for which the wrapping had to be adjusted (see _compat.py:358) - testing this proved to be hard as one would have to create a legacy module loader which provenly works - but I couldn't. So here, the testing is internal only...
  - while user code will still execute, PyCharm and other IDEs will likely not recognize the deprecated module references - the deprecated submodules won't show up in code completion and will look like they don't exist. 
 
Beyond maintainer review, I will also get some python experts to look at this PR and provide feedback to get a bit more confidence that this is not completely unrecommended.

TODO: 
 - [x] fix CI (coverage needs to be worked out with the subprocess separation)
 - [x] split out improvements to the `cirq.testing.assert_deprecated` and `_warn_or_error` method to a separate PR (done #3919)
 - [x] get python expert review

### CODE REVIEW

**Comment by balopat** [comment]
cc @cduck would love to have your opinion!

File: cirq/testing/_compat_test_data/__init__.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+# simulates a move of this module
+# fake_c -> module_a.module_b.module_c
+# but this won't create an attribute!
+_compat.deprecated_submodule(
+    new_module_name=f"{__name__}.module_a.module_b.module_c",
+    old_parent=__name__,
+    old_child="fake_c",
+    deadline="v0.20",
+    create_attribute=True,
+)
```
Comment (cduck) [inline cirq/testing/_compat_test_data/__init__.py:62]: You don't test `create_attribute=False`.

File: cirq/testing/_compat_test_data/module_a/__init__.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+from logging import info
+
+from cirq.testing._compat_test_data.module_a import (
+    module_b,
+)
+
+from cirq.testing._compat_test_data.module_a.dupe import (
+    DUPE_CONSTANT,
+)
+
```
Comment (cduck) [inline cirq/testing/_compat_test_data/module_a/__init__.py:10]: I would add a line like `test_value = 'module_a'` to all these files to use in the test assertions.

File: cirq/_compat_test.py [hunk 3]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
...
+
+
+def _import_deprecated_first_new_second():
+    """To ensure that module_a gets initialized only once."""
+    from cirq.testing._compat_test_data.fake_a import module_b
+
+    assert module_b
+    from cirq.testing._compat_test_data.module_a import module_b
+
+    assert module_b
```
Comment (cduck) [inline cirq/_compat_test.py:300]: How does this enforce single-initialization?  Same for the following several tests.

File: cirq/_compat_test.py [hunk 4]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
...
+    assert api
+
+
+# this is where the deprecation error should show where the deprecated usage
+# has occured, which is this file
+_deprecation_origin = ['_compat_test.py:']
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_a_deprecation_msg = [
+    'fake_a was used but is deprecated',
```
Comment (cduck) [inline cirq/_compat_test.py:391]: Should this be?
```suggestion
    'cirq.testing._compat_test_data.fake_a was used but is deprecated',
```

File: cirq/_compat_test.py [hunk 5]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
...
+
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_freezegun_deprecation_msg = [
+    'fake_freezegun was used but is deprecated',
+    'Use freezegun instead',
+] + _deprecation_origin
+
+
+def _trace_unhandled_exceptions(*args, **kwargs):
```
Comment (cduck) [inline cirq/_compat_test.py:415]: Nit: Simplify the kwargs code.
```suggestion
def _trace_unhandled_exceptions(*args, *, q, func, **kwargs):
```

File: cirq/_compat_test.py [hunk 6]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
...
+        # coverage: ignore
+        msg = str(ex)
+        queue.put((type(ex).__name__, msg, traceback.format_exc()))
+
+
+def subprocess_context(test_func):
+    """Ensures that sys.modules changes in subprocesses won't impact the parent process."""
+    import os
+
+    ctx = multiprocessing.get_context("spawn" if os.name == 'nt' else "fork")
```
Comment (cduck) [inline cirq/_compat_test.py:436]: Did you test on Windows?

Nit: Double quotes vs. single quotes.

File: cirq/_compat_test.py [hunk 7]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
...
+
+# see cirq_compat_test_data/__init__.py for the setup code
+_fake_freezegun_deprecation_msg = [
+    'fake_freezegun was used but is deprecated',
+    'Use freezegun instead',
+] + _deprecation_origin
+
+
+def _trace_unhandled_exceptions(*args, **kwargs):
+    from multiprocessing import Queue
```
Comment (cduck) [inline cirq/_compat_test.py:416]: Shouldn't this be in the global TYPE_CHECKING if statement at the top of the file?

File: cirq/_compat_test.py [hunk 8]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
...
+    queue: Queue = kwargs['q']
+    func = kwargs['func']
+    del kwargs['q']
+    del kwargs['func']
+    try:
+        func(*args, **kwargs)
+        queue.put(None)
+    # coverage: ignore
+    except BaseException as ex:
+        # coverage: ignore
```
Comment (cduck) [inline cirq/_compat_test.py:427]: Nit: Double ignore?

File: cirq/_compat_test.py [hunk 9]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
...
+        (_import_multiple_deprecated, [_fake_a_deprecation_msg, _fake_b_deprecation_msg]),
+        (_new_module_in_different_parent, [_fake_ops_deprecation_msg]),
+        # ignore the frame requirement - as we are using find_spec from importlib, it
+        # is detected as an "internal" frame by warnings
+        (_find_spec_deprecated_multiple_times, [_fake_a_deprecation_msg[:-1]]),
+        (_import_parent_use_constant_from_deprecated, [_fake_a_deprecation_msg]),
+        (_import_deprecated_sub_use_constant, [_fake_a_deprecation_msg]),
+        (_import_deprecated_same_name_in_earlier_subtree, [_fake_a_deprecation_msg]),
+        (_import_top_level_deprecated, [_fake_freezegun_deprecation_msg]),
+        (_from_deprecated_import_sub_of_sub, [_fake_a_deprecation_msg]),
```
Comment (cduck) [inline cirq/_compat_test.py:472]: It may be worth adding a test that just throws an error and is marked for pytest as expected to fail.  I think this would cover the above two `coverage: ignore`s.

File: cirq/_compat_test.py [hunk 10]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
...
+    # ensure that both packages are initialized exactly once
+    import cirq
+
+    with cirq.testing.assert_logs(
+        "init:compat_test_data",
+        "init:module_a",
+        min_level=logging.INFO,
+        max_level=logging.INFO,
+        count=2,
+    ):
```
Comment (cduck) [inline cirq/_compat_test.py:489]: Does nesting two `assert_logs` not work?

File: cirq/_compat_test.py [hunk 11]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
...
+def test_invalidate_caches():
+    called = False
+
+    class FakeFinder(importlib.abc.MetaPathFinder):
+        def invalidate_caches(self) -> None:
+            nonlocal called
+            called = True
+
+    DeprecatedModuleFinder(FakeFinder(), "new", None, "old", "v0.1").invalidate_caches()
+    assert called
```
Comment (cduck) [inline cirq/_compat_test.py:608]: Corner case you should also test:
- `cirq_google/file.py` tries to `import cirq.google.something_else`.  (Ideally this will never happen but you probably want to print a sensible error or warning to help developers.)

**Review by cduck — CHANGES_REQUESTED** [review]
This looks solid (as far as hacking the import machinery can be).  My comment on line 501 is my biggest suggestion.  The rest of my comments are just about robustness or improving test coverage.

File: dev_tools/conf/pip-install-minimal-for-pytest-changed-files.sh [hunk 12]
```diff
@@ -28,6 +28,7 @@ cat dev_tools/conf/pip-list-dev-tools.txt | grep pytest | xargs pip install
 cat dev_tools/conf/pip-list-dev-tools.txt | grep filelock | xargs pip install
 cat dev_tools/conf/pip-list-dev-tools.txt | grep freezegun | xargs pip install
 cat dev_tools/conf/pip-list-dev-tools.txt | grep flynt | xargs pip install
+pip install "importlib-metadata; python_version < '3.8'"
```
[the remark this answers is not in the document]
Comment (balopat) [reply dev_tools/conf/pip-install-minimal-for-pytest-changed-files.sh:31]: I tried, it doesn't really work with xargs for some reason, but let me give it another go

File: dev_tools/conf/mypy.ini [hunk 13]
```diff
@@ -5,7 +5,7 @@ follow_imports = silent
 ignore_missing_imports = true
 
 # 3rd-party libs for which we don't have stubs
-[mypy-apiclient.*,freezegun.*,matplotlib.*,mpl_toolkits,multiprocessing.dummy,numpy.*,oauth2client.*,pandas.*,pytest.*,scipy.*,sortedcontainers.*,setuptools.*,pylatex.*,networkx.*,qiskit.*,pypandoc.*,ply.*,_pytest.*,google.api.*,google.api_core.*,grpc.*,google.oauth2.*,google.protobuf.text_format.*,quimb.*,pyquil.*,google.cloud.*,filelock.*,codeowners.*,tqdm.*]
+[mypy-apiclient.*,freezegun.*,matplotlib.*,mpl_toolkits,multiprocessing.dummy,numpy.*,oauth2client.*,pandas.*,pytest.*,scipy.*,sortedcontainers.*,setuptools.*,pylatex.*,networkx.*,qiskit.*,pypandoc.*,ply.*,_pytest.*,google.api.*,google.api_core.*,grpc.*,google.oauth2.*,google.protobuf.text_format.*,quimb.*,pyquil.*,google.cloud.*,filelock.*,codeowners.*,tqdm.*,cirq.testing._compat_test_data.*,importlib_metadata.*]
```
[the remark this answers is not in the document]
Comment (balopat) [reply dev_tools/conf/mypy.ini:8]: it does! :)

File: cirq/_compat_test.py (hunk 3 above)
Comment (balopat) [reply cirq/_compat_test.py:300]: Oooops, I think what happened sometimes is that I pressed organize imports in PyCharm, that optimizes away unnecessary imports. I'll go back and fix all these. Thanks!

File: cirq/_compat_test.py (hunk 4 above)
Comment (balopat) [reply cirq/_compat_test.py:391]: It could be, it's just a partial matching, so it made it less verbose

File: cirq/_compat_test.py (hunk 6 above)
Comment (balopat) [reply cirq/_compat_test.py:436]: Yes, without this it was failing on Windows (which we do have in our test matrix).

File: cirq/_compat_test.py (hunk 10 above)
Comment (balopat) [reply cirq/_compat_test.py:489]: No, it doesn't, both would receive all the lines of logs.

**Comment by balopat** [comment]
Thank you so much for the thorough review @cduck - awesome points all around!

File: cirq/_compat.py [hunk 14]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+                sys.modules[self.old_module_name] = sys.modules[self.new_module_name]
+                return
+
+            # now we know we have to initialize the module
+            sys.modules[self.old_module_name] = module
+            sys.modules[self.new_module_name] = module
+
+            try:
+                return method(module)
+            except Exception as ex:
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:354]: 👍

File: cirq/_compat.py [hunk 15]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+                # if there's an error, we atomically remove both
+                del sys.modules[self.new_module_name]
+                del sys.modules[self.old_module_name]
+                raise ex
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:358]: 👍

File: cirq/_compat.py [hunk 16]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+
+        return exec_module
+
+
+def _is_internal(filename: str) -> bool:
+    """Returns whether filename is internal to python.
+
+    This is similar to how the built-in warnings module differentiates frames from internal modules.
+    """
+    return 'importlib' in filename and '_bootstrap' in filename
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:368]: They still have it in `https://github.com/python/cpython/blob/master/Lib/warnings.py#L275`

File: cirq/_compat.py [hunk 17]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+        new_module_name: the new module's fully qualified name
+    """
+
+    def __init__(self, loader: Any, old_module_name: str, new_module_name: str):
+        """A module loader that uses an existing module loader and intercepts
+        the execution of a module.
+        """
+        self.loader = loader
+        if hasattr(loader, 'exec_module'):
+            setattr(self, 'exec_module', self._wrap_exec_module(loader.exec_module))
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:322]: That's how I started, but mypy didn't like it, hence the setattr...

File: cirq/_compat.py [hunk 18]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+
+            def invalidate_caches() -> None:
+                return self.finder.invalidate_caches()
+
+            setattr(self, "invalidate_caches", invalidate_caches)
+
+    def find_spec(self, fullname: str, path: Any = None, target: Any = None) -> Any:
+        """Finds the specification of a module.
+
+        This is an implementation of the
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:436]: 👍

File: cirq/_compat.py [hunk 19]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+        Args:
+            fullname: name of the module
+            path: if presented, this is the parent module's submodule search path
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:439]: 👍

File: cirq/_compat.py [hunk 20]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+        """
+        if not fullname.startswith(self.old_module_name):
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:441]: Oh, good point, I'll add a test case 👍

File: cirq/_compat.py [hunk 21]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+            # if we are not interested in it, then just pass through to the wrapped finder
+            return self.finder.find_spec(fullname, path, target)
+
+        # warn for deprecation
+        _deduped_module_warn_or_error(self.old_module_name, self.new_module_name, self.deadline)
+
+        new_fullname = fullname.replace(self.old_module_name, self.new_module_name)
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:448]: I'll add a test case 👍

File: cirq/_compat.py [hunk 22]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+            )
+
+        # if the spec exists, return the DeprecatedModuleLoader that will do the loading as well
+        # as set the alias(es) in sys.modules as necessary
+        if spec is not None:
+            # change back the name to the deprecated module name
+            spec.name = fullname
+            # some loaders do a check to ensure the module's name is the same
+            # as the loader was created for
+            if hasattr(spec.loader, "name") and getattr(spec.loader, "name") == new_fullname:
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:468]: 👍

File: cirq/_compat.py [hunk 23]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+                new_fullname,
+                path=path,
+                target=target,
+            )
+
+        # if the spec exists, return the DeprecatedModuleLoader that will do the loading as well
+        # as set the alias(es) in sys.modules as necessary
+        if spec is not None:
+            # change back the name to the deprecated module name
+            spec.name = fullname
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:465]: Great question, let me check that.

File: cirq/_compat.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (cduck) [reply cirq/_compat.py:368]: OK.  Maybe mention in the comment that this is CPython specific.

File: cirq/_compat.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (cduck) [reply cirq/_compat.py:465]: Would be great to add a test for this too.

File: cirq/_compat_test.py (hunk 4 above)
Comment (cduck) [reply cirq/_compat_test.py:391]: Sure, but in this case I think it's an important part of the test.

File: cirq/_compat_test.py (hunk 5 above)
Comment (balopat) [reply cirq/_compat_test.py:415]: 👍

File: cirq/_compat.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:368]: 👍

File: cirq/_compat.py [hunk 24]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+            old_parent module
+    Returns:
+        None
+    """
+    _validate_deadline(deadline)
+
+    old_module_name = f"{old_parent}.{old_child}"
+
+    new_module_spec = importlib.util.find_spec(new_module_name)
+    new_module = importlib.import_module(new_module_name)
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:502]: +1 for only importing when create_attribute is True.

I'm not sure I follow fully your aside - the use case you mentioned will still work (as long as `create_attribute=True`) with the current setup as it's only one layer down and is tested in `_import_parent_use_constant_from_deprecated`.

[the remark this answers is not in the document]
Comment (cduck) [reply cirq/_compat.py:502]: The goal eventually is to separate cirq_google from cirq so people don't have to install cirq_google right?

All I ask for this PR is tests for `create_attribute=False` or remove the option for now.

File: cirq/_compat.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:448]: This made my head hurt a little bit.

File: cirq/_compat.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:465]: added (will commit soon) - the repr will be the new module's path in both cases.

File: cirq/_compat_test.py [hunk 25]
```diff
@@ -263,3 +270,339 @@ class BadlyDeprecatedClass(NewClass):
             ...
 
         # pylint: enable=unused-variable
+
+
+def _from_parent_import_deprecated():
+    from cirq.testing._compat_test_data import fake_a
+
+    assert fake_a
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat_test.py:278]: Sounds good.

File: dev_tools/conf/mypy.ini (hunk 13 above)
[the remark this answers is not in the document]
Comment (balopat) [reply dev_tools/conf/mypy.ini:8]: Well to be more precise it complains about all the fake references. Maybe it would be better to put the mypy exclusion closer to those lines?

File: cirq/testing/_compat_test_data/__init__.py (hunk 1 above)
Comment (balopat) [reply cirq/testing/_compat_test_data/__init__.py:62]: I added tests now.

File: cirq/testing/_compat_test_data/module_a/__init__.py (hunk 2 above)
Comment (balopat) [reply cirq/testing/_compat_test_data/module_a/__init__.py:10]: 👍 Added.

File: cirq/_compat_test.py (hunk 3 above)
Comment (balopat) [reply cirq/_compat_test.py:300]: Now that I'm looking at it - it does the job well, but probably could use an explanatory comment.

The way it works is through enforcing the count of info logs in `_test_deprecated_module_inner`: 

```
 with cirq.testing.assert_logs(
        "init:compat_test_data",
        "init:module_a",
        min_level=logging.INFO,
        max_level=logging.INFO,
        count=2,
    ):
```

It forces setting up the loader on the spec. Let's comment out this line in the `find_spec` method: 

```
spec.loader = DeprecatedModuleLoader(spec.loader, fullname, new_fullname)
```
In case of the following user code

```
from cirq.testing._compat_test_data.fake_a import module_b
from cirq.testing._compat_test_data.module_a import module_b
```

The importlib machinery first finds `fake_a` with our `DeprecatedModuleFinder` as `module_a`, it sets up a default Loader, which loads the module and places is the reference in `sys.modules['fake_a']`. 
Then, the second line executes again, finds `module_a` (which is NOT in the module cache at the moment), sets up a default Loader, which loads the module again, and places the reference in `sys.modules['module_a']`. 

The assertion error I get is 

```
 AssertionError: Expected 2 log message but got 3. Log messages: ['init:compat_test_data', 'init:module_a', 'init:module_a']
```

This is the double execution problem with these references. The `DeprecatedModuleLoader` does the "atomic" (not really as it's not prepared for multithreading...) update of the two references.

File: cirq/_compat_test.py (hunk 4 above)
Comment (balopat) [reply cirq/_compat_test.py:391]: 👍 added

File: cirq/_compat_test.py (hunk 7 above)
Comment (balopat) [reply cirq/_compat_test.py:416]: actually we do import multiprocessing already, I just simplified it in the function argument as a type annotation

File: cirq/_compat_test.py (hunk 9 above)
Comment (balopat) [reply cirq/_compat_test.py:472]: 👍 great idea. Added.

File: cirq/_compat_test.py (hunk 8 above)
Comment (balopat) [reply cirq/_compat_test.py:427]: solved by adding a test for it.

File: dev_tools/conf/mypy.ini (hunk 13 above)
[the remark this answers is not in the document]
Comment (cduck) [reply dev_tools/conf/mypy.ini:8]: So mypy is complaining about the `import ...fake_a` lines in the tests?  I'd just tell mypy to ignore those lines specifically.

File: cirq/_compat.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:502]: Got it! Yes, that is partially the goal - but I imagine this functionality to cater for the generic case of deprecating modules. Which means that in cases where we don't import submodules as an attribute this should still work. 

I moved importing the module inside the `if create_attribute:` block. I also modified all the setups to be create_attribute=False to demonstrate that it wasn't required for any of the use cases, except the `module_a` one. I also added an explicit test to see `create_attribute=False` doesn't create the attribute.

File: cirq/_compat.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (cduck) [reply cirq/_compat.py:322]: Your choice.  Mypy isn't much help either way for this kind of hacking.

File: cirq/_compat.py [hunk 26]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+
+    class Wrapped(ModuleType):
+        __dict__ = parent_module.__dict__
+
+        def __getattr__(self, name):
+            if name == old_child:
+                _deduped_module_warn_or_error(
+                    f"{old_parent}.{old_child}", new_module_name, deadline
+                )
+            return getattr(parent_module, name)
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:535]: `ModuleType` does not specify `__getattr__`, it relies purely on `__dict__`, thus `super().__getattr__` doesn't exist. However, I noticed that `__dict__` was not necessary at all.

File: cirq/_compat_test.py (hunk 11 above)
Comment (balopat) [reply cirq/_compat_test.py:608]: For the first one - as soon as that happens in a pytest, we'll throw an error - we are prohibiting usage of deprecated features from Cirq.

For the second - the `repeated` use cases that I am adding now cover both the old and new access patterns work with a path that has a substring equal to the deprecated module name.

File: cirq/_compat.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:322]: Yeah, those are the two recommended options: `# type: ignore` or `setattr` - I find actually the former one more readable so I'll go with that. https://github.com/python/mypy/issues/2427

File: cirq/_compat.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:475]: This is done. I think having the new module `cirq_google.MyClass` should be okay, but let me know if you have thoughts on this.

File: cirq/_compat.py [hunk 27]
```diff
@@ -285,3 +298,240 @@ def __getattr__(self, name):
...
+            the submodules of the deprecated module
+        old_module_name: the deprecated module's fully qualified name
+        deadline: the deprecation deadline
+    """
+
+    def __init__(
+        self,
+        finder: Any,
+        new_module_name: str,
+        new_module_spec: ModuleSpec,
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:410]: You found the one thing I am not exactly sure about :) 
I'll have to dig into this a bit. If I just let the Finder find the spec again, I have 3 failures: 

```
FAILED cirq/_compat_test.py::test_deprecated_module[_import_multiple_deprecated-deprecation_messages5] - Failed: ModuleNotFoundError: No module named 'cirq.testing._compat_test_data.fake_b'
FAILED cirq/_compat_test.py::test_deprecated_module[_new_module_in_different_parent-deprecation_messages6] - Failed: ModuleNotFoundError: No module named 'cirq.testing._compat_test_data.fake_ops'
FAILED cirq/_compat_test.py::test_deprecated_module[_import_top_level_deprecated-deprecation_messages11] - Failed: ModuleNotFoundError: No module named 'cirq.testing._compat_test_data.fake_freezegun'
=
```

Let me think this through, most of the other comments are addressed.

**Comment by cduck** [comment]
@balopat You need `xargs -0`.

**Comment by cduck** [comment]
`... | tr \\n \\0 | xargs -0 pip install` in case you have multiple matching lines.

File: cirq/_compat.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (cduck) [reply cirq/_compat.py:475]: I prefer `cirq_google.MyClass`.  I wouldn't want it to change when you eventually remove the deprecated module.

File: cirq/_compat.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (cduck) [reply cirq/_compat.py:410]: I suspect this is because the `path` (and `target`) arguments to `finder.find_spec` need to be adjusted for the new parent module (https://docs.python.org/3/library/importlib.html#importlib.abc.MetaPathFinder.find_spec).

Try
```python
        # find the corresponding spec in the new structure
        if fullname == self.old_module_name:
            path = None
            target = None
        spec = self.finder.find_spec(
            new_fullname,
            path=path,
            target=target,
        )
```

File: cirq/_compat.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:475]: 👍

**Comment by balopat** [comment]
> @balopat You need `xargs -0`: https://unix.stackexchange.com/questions/38148/why-does-xargs-strip-quotes-from-input

Oh yes, of course :) Thanks for the hint.

File: cirq/_compat.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:410]: You are right, this does boil down to updating path - and maybe target, though I haven't seen any Finder yet that uses target.

E.g. this works:

```
      if fullname == self.old_module_name:
            spec = self.finder.find_spec(
                new_fullname,
                path=["/Users/balintp/dev/proj/Cirq/cirq/testing/_compat_test_data/module_a",
                      "/Users/balintp/dev/proj/Cirq/cirq/testing/_compat_test_data",
                      "/Users/balintp/dev/proj/Cirq/cirq",
                      "/Users/balintp/dev/proj/Cirq/cirq/testing",
                      ] + sys.path,
                target=None,
            )
```

However, finding the parent module's path is a harder problem than it looks. 
Finders are not smart enough to be able to figure the parent path by themselves. 

Based on https://github.com/python/cpython/blob/master/Lib/importlib/util.py#L92-L103 this also works:


```
        # find the corresponding spec in the new structure
        if fullname == self.old_module_name:
            # this is the first time the deprecated module is being found
            # which means that the new parent needs to be found first and under
            # the new parent's path, we should be able to find the new name of
            # the deprecated module
            # this code is heavily inspired by importlib.util.find_spec
            parent_name = new_fullname.rpartition('.')[0]
            if parent_name:
                parent = __import__(parent_name, fromlist=['__path__'])
                # note that compared to importlib.util.find_spec we don't handle
                # AttributeError here because it is not expected to happen in case
                # of a DeprecatedModuleLoader - the new parent should exist and be
                # a proper package
                parent_path = parent.__path__
            else:
                parent_path = None
            spec = self.finder.find_spec(new_fullname, parent_path, None)
        else:
            # we are finding a submodule of the parent of the deprecated module,
            # which means that the parent was already found, and thus, `path` is
            # correctly pointing to the module's parent in the new hierarchy
            spec = self.finder.find_spec(
                new_fullname,
                path=path,
                target=target,
            )
```

It's a bit more complicated, but it is cleaner, more principled. I'll update it :)

File: dev_tools/conf/mypy.ini (hunk 13 above)
[the remark this answers is not in the document]
Comment (balopat) [reply dev_tools/conf/mypy.ini]: Done!

**Comment by balopat** [comment]
Ready for another look @cduck, again, thank you for your input 🙇 !!!

File: cirq/_compat.py [hunk 28]
```diff
@@ -459,14 +455,36 @@ def find_spec(self, fullname: str, path: Any = None, target: Any = None) -> Any:
...
+            if parent_name:
+                parent = __import__(parent_name, fromlist=['__path__'])
+                # note that compared to importlib.util.find_spec we don't handle
+                # AttributeError here because it is not expected to happen in case
+                # of a DeprecatedModuleLoader - the new parent should exist and be
+                # a proper package
+                parent_path = parent.__path__
+            else:
+                parent_path = None
+            spec = self.finder.find_spec(new_fullname, parent_path, None)
```
Comment (cduck) [inline cirq/_compat.py:473]: Thanks for looking into this.  I didn't realize it would get this involved.  Now I see the core challenge here with finding the spec while bypassing the wrapper.  This solution works.

**Review by cduck — APPROVED** [review]
Wow, that was a lot of double quotes.  I stopped noticing them after the first few.

LGTM after cleaning up some debugging code.

File: cirq/_compat.py [hunk 29]
```diff
@@ -459,14 +455,36 @@ def find_spec(self, fullname: str, path: Any = None, target: Any = None) -> Any:
...
             spec = self.finder.find_spec(
                 new_fullname,
                 path=path,
                 target=target,
             )
+        f = self
+        finders = f"{f} > "
+        while isinstance(f, DeprecatedModuleFinder):
+            f = f.finder
+            finders += f" > {f}"
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:487]: yupp, thanks for spotting that, these were both for debugging purposes.

File: cirq/_compat.py [hunk 30]
```diff
@@ -480,6 +498,9 @@ def find_spec(self, fullname: str, path: Any = None, target: Any = None) -> Any:
             spec.loader = DeprecatedModuleLoader(spec.loader, fullname, new_fullname)
         return spec
 
+    def __str__(self) -> str:
+        return f"DMF({id(self)} - {self.old_module_name} -> {self.new_module_name})"
+
```
[the remark this answers is not in the document]
Comment (balopat) [reply cirq/_compat.py:503]: yupp, thanks for spotting that, these were both for debugging purposes.

File: cirq/_compat.py (hunk 28 above)
Comment (balopat) [reply cirq/_compat.py:473]: Awesome! Thank you for the detailed, thorough review.

**Comment by HaoZeke** [comment]
I was wondering if the design doc for this could be made public @balopat.

**Comment by dabacon** [comment]
cc @MichaelBroughton

**Comment by tanujkhattar** [comment]
@HaoZeke The design document linked in the original PR description is open to all cirq developers. You can join the cirq-dev group (open to public) to view the design doc. 

Join Google Group: https://groups.google.com/g/cirq-dev
Design Doc Link: http://tinyurl.com/cirq-deep-aliasing

**Comment by HaoZeke** [comment]
Fantastic, thanks!
